File: Java/PrisonerSimulator/src/Player/PlayerFactory.java
package Player;

import Player.Strategy.IStrategy;
import Player.Strategy.KindRepeaterStrategy;

import java.util.InputMismatchException;
import java.util.Objects;

public class PlayerFactory {
    private int idCounter = 0;

    private IStrategy createStrategyByName(String name) {
        if (Objects.equals(name, "KindRepeaterStrategy")) {
            return new KindRepeaterStrategy();
        }
        // TODO: Implement rest of strategies

        throw new InputMismatchException(String.format("Unknown strategy name: %s", name));
    }

    public Player createPlayer(String strategy_name) {
        var strategy = createStrategyByName(strategy_name);
        int id = idCounter++;
        var name = strategy_name + id;

        return new Player(id, name, strategy);
    }
}

File: Kotlin/PrisonerSimulator/src/player/PlayerFactory.kt
package player

import player.strategy.*

class PlayerFactory {
    private var idCounter = 0

    fun createPlayer(strategyName: String): Player {
        val strategy = when (strategyName) {
            "EvilDumbassStrategy" -> EvilDumbassStrategy()
            "EvilRepeaterStrategy" -> EvilRepeaterStrategy()
            "InconsistentStrategy" -> InconsistentStrategy()
            "KindDumbassStrategy" -> KindDumbassStrategy()
            "KindForgivingRepeaterStrategy" -> KindForgivingRepeaterStrategy()
            "KindRepeaterStrategy" -> KindRepeaterStrategy()
            "RandomStrategy" -> RandomStrategy()
            else -> throw IllegalArgumentException("Unknown strategy name: $strategyName")
        }

        return Player(++idCounter, strategyName + idCounter, strategy)
    }
}
